qwerty_ast/src/typecheck.rs

```diff
@@ -770,6 +860,128 @@ fn typecheck_basis(basis: &Basis, env: &mut TypeEnv) -> Result<Type, TypeError>
     }
 }
 
+/// Determines if an expression is inherently reversible.
+fn is_expr_inherently_reversible(expr: &Expr, env: &mut TypeEnv) -> Result<bool, TypeError> {
+    match expr {
+        // Operations that break reversibility
+        Expr::Measure { .. } => Ok(false),
+        Expr::Discard { .. } => Ok(false),
+        Expr::Conditional { .. } => Ok(false), // Classical conditionals (if/else) are not allowed in reversible functions
+
+        Expr::Pipe { lhs, rhs, .. } => {
+            let lhs_is_rev = is_expr_inherently_reversible(lhs, env)?;
+            let rhs_is_rev = is_expr_inherently_reversible(rhs, env)?;
+            if !lhs_is_rev || !rhs_is_rev {
+                return Ok(false);
+            }
+
+            let rhs_ty = typecheck_expr(rhs, env)?; // Recursively typecheck to get the actual type of RHS
```

Comment:
Am I correct that this would throw a "variable not found" error for the last line below because `env` is not being populated with `Assign`s as `infer_function_body_reversibility()` iterates through the statements in the body?

```python
@qpu
@reversible
def trolling(q: qubit) -> qubit:
    m = measure
    return '0' | m
```

(In AST terms, `@reversible` means `is_rev=true` on the `FunctionDef`; the `=` is an `Assign` statement; and the `m` on the last line is a `Variable` expression)

Reply:
I have an idea below in a comment on `infer_function_body_reversibility()` that could help with this